File: src/core/crypto/token_manager.py
import hmac
import hashlib
from datetime import datetime, timezone, timedelta
from typing import Tuple, Optional
from uuid import UUID

from src.config import get_settings


class TokenManager:
    def __init__(self, secret_key: str, max_ttl_seconds: int = 900):
        self.secret_key = secret_key.encode("utf-8")
        self.max_ttl_seconds = max_ttl_seconds

    def create_session_token(
        self, certificate_id: UUID, certificate_expiry: datetime
    ) -> Tuple[str, int]:
        """Create short-lived session token with real-time expiry bound"""
        now = datetime.now(timezone.utc)
        remaining = (certificate_expiry - now).total_seconds()

        # Token TTL = min(max_ttl, remaining certificate lifetime)
        ttl_seconds = max(1, min(self.max_ttl_seconds, remaining))
        token_expiry = now + timedelta(seconds=ttl_seconds)

        payload = f"{certificate_id}|{token_expiry.isoformat().replace('+', '_')}"
        signature = hmac.new(
            self.secret_key, payload.encode("utf-8"), hashlib.sha256
        ).hexdigest()

        token = f"{payload}|{signature}"
        return token, int(ttl_seconds)

    def validate_session_token(
        self, token: str
    ) -> Tuple[Optional[UUID], Optional[datetime], bool]:
        """Validate session token and return certificate_id and expiry"""
        try:
            parts = token.split("|")
            if len(parts) != 3:
                return None, None, False

            cert_id_str, expiry_iso, signature = parts
            token_expiry = datetime.fromisoformat(expiry_iso.replace('_', '+'))

            # Verify signature
            payload = f"{cert_id_str}|{expiry_iso}"
            expected = hmac.new(
                self.secret_key, payload.encode("utf-8"), hashlib.sha256
            ).hexdigest()

            if not hmac.compare_digest(expected, signature):
                return None, None, False

            # Check if token itself is expired
            if token_expiry < datetime.now(timezone.utc):
                return None, None, False

            return UUID(cert_id_str), token_expiry, True

        except Exception:
            return None, None, False


def get_token_manager() -> TokenManager:
    return TokenManager(get_settings().secret_key, get_settings().token_ttl_seconds)

File: app/Services/Crypto/TokenManager.php
<?php

namespace App\Services\Crypto;

use DateTimeImmutable;
use DateTimeZone;
use Exception;
use InvalidArgumentException;

class TokenManager
{
    protected string $secretKey;
    protected int $maxTtlSeconds;

    public function __construct(string $secretKey, int $maxTtlSeconds = 900)
    {
        if (empty($secretKey)) {
            throw new InvalidArgumentException("Secret key cannot be empty.");
        }
        $this->secretKey = $secretKey;
        $this->maxTtlSeconds = $maxTtlSeconds;
    }

    /**
     * Create short-lived session token with real-time expiry bound.
     * * @return array{0: string, 1: int} Returns [token, ttl_seconds]
     */
    public function createSessionToken(string $certificateId, DateTimeImmutable $certificateExpiry): array
    {
        $now = new DateTimeImmutable('now', new DateTimeZone('UTC'));

        // Calculate remaining certificate lifetime in seconds
        $remaining = $certificateExpiry->getTimestamp() - $now->getTimestamp();

        // Token TTL = min(max_ttl, remaining certificate lifetime). Ensure it's at least 1.
        $ttlSeconds = max(1, min($this->maxTtlSeconds, $remaining));
        $tokenExpiry = $now->modify("+{$ttlSeconds} seconds");

        // Format ISO string, replacing '+' with '_' to match your Python token string syntax perfectly
        $expiryIso = str_replace('+', '_', $tokenExpiry->format(DateTimeImmutable::ATOM));

        $payload = "{$certificateId}|{$expiryIso}";
        $signature = hash_hmac('sha256', $payload, $this->secretKey);

        $token = "{$payload}|{$signature}";

        return [$token, $ttlSeconds];
    }

    /**
     * Validate session token and extract parameters.
     * * @return array{0: ?string, 1: ?DateTimeImmutable, 2: bool} Returns [certificate_id, token_expiry, is_valid]
     */
    public function validateSessionToken(string $token): array
    {
        try {
            $parts = explode('|', $token);
            if (count($parts) !== 3) {
                return [null, null, false];
            }

            [$certificateId, $expiryIso, $signature] = $parts;

            // Revert token-safe formatting back to proper ISO format
            $normalizedIso = str_replace('_', '+', $expiryIso);
            $tokenExpiry = new DateTimeImmutable($normalizedIso, new DateTimeZone('UTC'));

            // Verify HMAC signature
            $payload = "{$certificateId}|{$expiryIso}";
            $expectedSignature = hash_hmac('sha256', $payload, $this->secretKey);

            if (!hash_equals($expectedSignature, $signature)) {
                return [null, null, false];
            }

            // Check if the token itself has expired chronologically
            $now = new DateTimeImmutable('now', new DateTimeZone('UTC'));
            if ($tokenExpiry < $now) {
                return [null, null, false];
            }

            return [$certificateId, $tokenExpiry, true];

        } catch (Exception $e) {
            return [null, null, false];
        }
    }
}